### DISCUSSION TITLE
Thoughts on a Book AST - bast?

### DISCUSSION THREAD

**Discussion by aegatlin** [syntax-tree/unist#82] [Ideas] [answered]
I have a stale project at [libry.io](https://libry.io/) that I was writing in Elixir. I was not familiar with unified or AST's in general at that time. I later realized I was essentially building a book AST in json. I was parsing (public domain) books either from [project gutenberg](https://www.gutenberg.org/) or [standard ebooks](https://standardebooks.org/) and turning them into what I was calling at the time a `json_book`. Most of the nodes were similar to mdast. Some were similar to nlcst nodes, and fundamentally its value store was plaintext like nlcst, though my implementation was more abstract than concrete. It also introduced new nodes like Chapter and Section, and the root node had a Toc struct for table of contents.

When I discovered unified I got excited by the potential to leverage the ecosystem. I think a bast implementation could change the way digital books are experienced online, referenced, highlighted, etc. Transforming to and from bast from the other unist-compatible trees could be fruitful.

I'm curious what you think about a Book AST?

I'm surely naive in a myriad of ways, including how to get started. I'm afraid its too similar to some of the other ASTs and I wouldn't want to "duplicate" nodes that are already in some of the other trees, but at the same time, some nodes like `Chapter` as `Section` feel like the belong in a new tree system.

**Comment by ChristianMurphy** [comment, marked as answer]
Two ways of approaching it.

---

First approach is if you want to create a new community project, go for it!
Several other projects have:

* https://github.com/orgapp/orgajs#readme
* https://github.com/rasendubi/uniorg#readme
* https://github.com/phuctm97/ntast#readme
* https://github.com/redotjs/redot#readme

Are a few I know of off the top of my head I'm sure there are more.

As long as your AST implements/extends https://github.com/syntax-tree/unist#readme
It should work well with unifiedjs tools.

Once you have an AST spec, https://github.com/orgs/syntax-tree/discussions/36 offers some ideas of next steps.

---

Second approach if you want to work with an existing book format, for example EPUB.
There may be existing community tools which can help https://github.com/syntax-tree/xast comes to mind, as EPUB is XML.
And xast pairs well with the rest of the ecosystem, and already has some utilities, which you could use as building blocks to create any transformers, compilers, etc you want to.

**Reply by aegatlin** [reply]
Thank you for sharing these resources, I'm down the rabbit hole! 🙏
